test/integration/replicaset/replicaset_test.go

```diff
@@ -162,19 +173,252 @@ func rmSimpleSetup(t *testing.T) (*httptest.Server, framework.CloseFunc, clients
+	var createdPods []*v1.Pod
+	for _, rs := range rss {
+		createdRS, err := rsClient.Create(rs)
+		if err != nil {
+			t.Fatalf("Failed to create replica set %s: %v", rs.Name, err)
+		}
+		createdRSs = append(createdRSs, createdRS)
+	}
+	for _, pod := range pods {
+		createdPod, err := podClient.Create(pod)
+		if err != nil {
+			t.Fatalf("Failed to create pod %s: %v", pod.Name, err)
+		}
+		createdPods = append(createdPods, createdPod)
+	}
+
+	return createdRSs, createdPods
+}
+
+// Verify .Status.Replicas is equal to .Spec.Replicas
+func waitRSStable(t *testing.T, clientSet clientset.Interface, rs *v1beta1.ReplicaSet) {
+	rsClient := clientSet.Extensions().ReplicaSets(rs.Namespace)
+	if err := wait.PollImmediate(interval, timeout, func() (bool, error) {
+		newRS, err := rsClient.Get(rs.Name, metav1.GetOptions{})
+		if err != nil {
+			return false, err
+		}
+		return newRS.Status.Replicas == *rs.Spec.Replicas, nil
+	}); err != nil {
+		t.Fatalf("Failed to verify .Status.Replicas is equal to .Spec.Replicas for rs %s: %v", rs.Name, err)
+	}
+}
+
+// Update .Spec.Replicas to replicas and verify .Status.Replicas is changed accordingly
+func scaleRS(t *testing.T, c clientset.Interface, rs *v1beta1.ReplicaSet, replicas int32) {
+	rsClient := c.Extensions().ReplicaSets(rs.Namespace)
+	if err := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		newRS, err := rsClient.Get(rs.Name, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+		*newRS.Spec.Replicas = replicas
+		rs, err = rsClient.Update(newRS)
+		return err
+	}); err != nil {
+		t.Fatalf("Failed to update .Spec.Replicas to %d for rs %s: %v", replicas, rs.Name, err)
+	}
+	waitRSStable(t, c, rs)
+}
+
+func updatePod(t *testing.T, podClient typedv1.PodInterface, pod *v1.Pod, updateFunc func(*v1.Pod)) {
+	if err := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		newPod, err := podClient.Get(pod.Name, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+		updateFunc(newPod)
+		_, err = podClient.Update(newPod)
+		return err
+	}); err != nil {
+		t.Fatalf("Failed to update pod %s: %v", pod.Name, err)
+	}
+}
+
+func updatePodStatus(t *testing.T, podClient typedv1.PodInterface, pod *v1.Pod, updateStatusFunc func(*v1.Pod)) {
+	if err := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		newPod, err := podClient.Get(pod.Name, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+		updateStatusFunc(newPod)
+		_, err = podClient.UpdateStatus(newPod)
+		return err
+	}); err != nil {
+		t.Fatalf("Failed to update status of pod %s: %v", pod.Name, err)
+	}
+}
+
+func getPods(t *testing.T, podClient typedv1.PodInterface, labelMap map[string]string) *v1.PodList {
+	podSelector := labels.Set(labelMap).AsSelector()
+	options := metav1.ListOptions{LabelSelector: podSelector.String()}
+	pods, err := podClient.List(options)
+	if err != nil {
+		t.Fatalf("Failed obtaining a list of pods that match the pod labels %v: %v", labelMap, err)
+	}
+	if pods == nil {
```

Comment:
In general, if a Go function returns both a value and an `error`, you can assume that the return value is valid as long as the returned `error` is `nil`. For example, if the function ever returns `nil, nil` that means that `nil` is a valid value that the caller may expect to see, so you don't want to catch that here and prevent it from going through. If, on the other hand, `nil` is an invalid value for `pods`, then `List()` should return a non-nil `err` in that case, so again you are not responsible for checking for `pods == nil`.

Reply:
Thanks for pointing it out! I did think about this before, but was not sure about it. I will remove `pods == nil` check.